### ISSUE TITLE
LaTeX2e: character per character filename parsing done by \InputIfFileExists (et al.)  looks costly to an ignorant external observer/user

### ISSUE THREAD

**Issue by jfbu** [latex3/latex2e#523]
## Brief outline of the question

Yesterday I investigated a minted issue. The minted package caches files using two md5sums in the basename so the basename is 64 characters long. Examining with `\tracingmacros1` I discovered what appeared to be thousands of expl3 macro expansions during `\minted@input` which does `\IfFileExists` and then `\input`.

I have now tested with `\InputIfFileExists` using current TL 2021-pretest pdflatex (not pdflatex-dev).

For filenames without extension it appears there is a cost of 125 macro calls per character in the filename:

>  Law of input for extension less filenames: (414 + 125 * (length of filename)) macro calls

For filenames with extension this appears to drop to 75 macro calls per character.

> Law of input for filenames with extension: (245 + 75 * (length of filename inclusive of extension)) macro calls

I understand there is something impolite in asking ; Wittgenstein as a kid realized politeness in society meant lying and opted for a life out of the expected pathways. But, I wonder if the character per character approach does not have alternatives.

I vaguely understand the parsing is due to checking for quote character and space character and perhaps the dot character (I have not looked).

## Minimal example showing the bug

```latex
\RequirePackage{latexbug}       % <--should be always the first line (see CONTRIBUTING)!
\documentclass{article}

\newcommand\testinput[1]{%
\typeout{START TEST #1}%
  \tracingmacros1
  \tracingstacklevels2
  \InputIfFileExists{#1}{}
  \tracingmacros0
\typeout{STOP TEST #1}%
}
\begin{document}


% 1
\testinput{A}
% 539 macro calls

% 2
\testinput{AA}
% 664 macro calls

% 3
\testinput{AAA}
% 789 macro calls

% 4
\testinput{AAAA}
% 914 macro calls

% 8
\testinput{AAAAAAAA}
% 1414 macro calls

% 16
\testinput{AAAAAAAAAAAAAAAA}
% 2414 macro calls

% 64
\testinput{AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA}
% 8414 macro calls

Law of input for extension less filenames: (414 + 125 * (length of filename)) macro calls

% 1
\testinput{A.pygtex}
% 845 macro calls

% 2
\testinput{AA.pygtex}
% 920 macro calls

% 3
\testinput{AAA.pygtex}
% 995 macro calls

% 4
\testinput{AAAA.pygtex}
% 1070 macro calls

% 8
\testinput{AAAAAAAA.pygtex}
% 1370 macro calls

% 16
\testinput{AAAAAAAAAAAAAAAA.pygtex}
% 1970 macro calls

% 64
\testinput{AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA.pygtex}
% 5570 macro calls

Law of input for filenames with extension: (245 + 75 * (length of filename inclusive of extension)) macro calls

% check
% 16
\testinput{AAAAAAAAAAAAAAAA.abc}
% 1745 macro calls (= 245 + 75 * (16+1+3))

% 64
\testinput{AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA.abc}
% 5345 macro calls (= 245 + 75 * (64+1+3))


\end{document}
```

## Log file (required) and possibly PDF file  

[testinput.log](https://github.com/latex3/latex2e/files/6094960/testinput.log)

**Comment by PhelypeOleinik** [comment]
Thanks for the detailed report!

The character-by-character parsing is built into `expl3`'s `\__kernel_file_name_sanitize:n`, which does the quote-checking, space-preserving, it also does selective macro expansion (protected or robust macros are left alone, whereas everything else is expanded), and it avoids expanding active characters. At this stage there should be no differentiation on names with and without extensions (other than the length which adds more cycles to the process).

After that the code does some (that I expected to be) light processing to separate out the file name in path, base, and extension, so I'd expect _some_ difference, but not _that much_ difference...

I was already aware that the file name parsing grew a _lot_ with the 2020-10-01 release, so I'll use this issue to keep track of the progress.  I expect to improve the code on this area.

A slightly you-focussed question:  did you notice this issue due to the length of the trace or due to increased run time?

**Comment by jfbu** [comment]
> A slightly you-focussed question: did you notice this issue due to the length of the trace or due to increased run time?

Due to length of the trace. I was investigating some question involving minted package, whose code base is unfamiliar to me, involving some input operations. I thus added my usual `\tracingmacros1` and start looking in log. Hitting repeatedly in my Emacs C-v to let code pass by, one windows size at a time, and again, again, again, then I became curious it seemed very long: thousands and thousands of lines. So this morning I took the matter again with `\tracingstacklevels` to quantify the matter.

The increase in run time is certainly humanly negligible and I did not try to quantify. Doing this now with the above test file and all tracing operations removed, I see that all the combined `\InputIfFileExists` amount to less than 3 hundredths of a second at my locale (2.8Ghz), whereas with TeXLive 2019 (fully updated) it was a bit more than 0.5 hundredths of a second. So it definitely isn't like a 100X increase, more like 6X. No big deal in terms of time.

**Comment by PhelypeOleinik** [comment]
> No big deal in terms of time.

That's what I hoped.  I did some light testing locally before pushing those changes and they didn't seem like show-stoppers.

Regarding the length of the trace, all the fun of filename-parsing happens in `\set@curr@file`, within a group, so maybe it would be good to do `\tracingnone` inside that group, so that it doesn't weigh that much in the logs.

Which reminds me, we should add support to `\tracingstacklevels` and `\tracinglostchars=3` in `\tracingall/none`

**Comment by jfbu** [comment]
@PhelypeOleinik From a design point of view, although as I said the time penalty compared to previous LaTeX2e is only about 5X it seems (but I got this estimate by mixing usage with short and with long filenames), one wonders why a macro which will be used by all users and package authors should cater for such exotic things such as using `\sqrt` in the filename. Users need support for non-ascii chars,  spaces, and the filename (or some part of it) should be allowed to be abstracted in a "command": so far `\newcommand` user interface has been only a higher level `\long\def`, and same for "command" with mandatory arguments, however a LaTeX2e command with optional argument is more problematic and by the way I have no idea if LaTeX2e's  `\input` allows it because it never crossed my mind to try, in fact nothing else than a zero- or one-argument "command" I would have ever considered. 

Naturally, non-ascii characters which may have been handled by inputenc on non-Unicode engine should work nicely, without the user worrying. But this is a responsability of inputenc to have made provisions for some context telling it how it should go, and of course, naturally, inputenc defines its active bytes so that it can be told to behave nicely.

Then there is the case of active character such as `~`, or more generally the few characters known by LaTeX user to be special such as `#` or `$`. I see the three examples I give now are handled gracefully which was not the case in older LaTeX, such as from TL2016 and I don't know if the changes all came with 2020-10-01, or earlier, or now. For decades basic LaTeX user could not use them with `\input`, if not being told by some guru to use some non-LaTeXian  thing such as `\string` (which worked with `~` and `$`,  not with `#`). This works now without `\string`. But we still can't do `\texttt{\input{file#name}}` as it raises 

```
! Illegal parameter number in definition of \reserved@a.
```

My point here is that from the expandable core TeX `\input` with its characteristics in terms of special characters (they are special, so need special handling) and macros (well, they need to be expandable and expand to some "string") , to the actual LaTeX2e implementation which takes lots of macro expansion to provide a functionality unneeded in  vast majority of use cases (or to the contrary absolutely, definitely, is strongly needed such as allowing non-ascii letters; but should take only a few code lines via setting the suitable inputenc context), there was some intermediary which could be considered. My personal feeling is that having a very fast `\input` with documented historic limitations alongside a new `\safeinput` with documented new possibilities is better than adding substantial overload to the existing `\input` which pre-exists in thousands of packages and zillions of user document and so far did not need to handle such exotism as robust macros giving portions of the filenames. I feel it strange that suddenly all those user documents which worked, still do work, but do thousands of additional macro expansions.

Concretely to me, it means I will be tempted to only ever use  `\@@input` even in user-level documents because I understand what it does, I know its limitations and benefits, and that nothing can beat the primitive in terms of speed, and any higher level construct will at some point end up using the primitive after some malaxing which I can very well do myself in almost all use cases with some `\string` etc, the only thorny point being with non-ascii characters as I then would have to look into inputenc source code to learn what is correct way to tell it to let the active bytes behave nicely here, so as user I would adhere to "if you want non-ascii letters and spaces, use our nice `\input`;  if you need special characters such as `#` or `~`, use our `\safeinput`". Saying all this with some fear that this will re-inforce future decision to actually let the TeX primitive not even be usable at all in future...

Definitely the LaTeX3 programming language is well established, well motivated, and expertly crafted. But if it is to be used in all LaTeX aspects, then I do feel the whole LaTeX2e should be given a new name LaTeX3, and LaTeX2e, like TeX should have been given a quasi-final "bugs are features" stage.

**Comment by PhelypeOleinik** [comment]
@jfbu We decided to change how the file name is parsed to a simpler `\csname`-based approach (latex3/latex3#869). Next time `l3kernel` is updated, the filename parsing routine should be much shorter, with more or less the same feature set (except that as usual a `\protected` macro is expanded in a `\csname`, whereas it was not in the token-by-token expansion code).  I tested your example and it does 335 macro expansions for a file without extension and 198 for a file with extension, regardless of name length (assuming, of course, no macros or utf-8 chars in the file name).

**Comment by FrankMittelbach** [comment]
> (assuming, of course, no macros or utf-8 chars in the file name).

it should not significantly change with utf8 chars in the csname approach: they have a very small number of expansions each if inside a csname and then result in  \string ut8-fchar (or rather the equiv in bytes).

**Comment by FrankMittelbach** [comment]
that one is actually premature  @PhelypeOleinik  :-) as it depends on the release of expl3 which is not yet out but will in a few days  --- I'll try to keep the status of issues open until code is actually available on CTAN

**Comment by PhelypeOleinik** [comment]
@FrankMittelbach Oops, indeed! Sorry (and I've just closed #528, before seeing your message).  I'll reopen
